devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,166 @@
+    schedstat_path = '/proc/schedstat'
+
+    def setup(self):
+        try:
+            self.old_sysctl_value = self.target.read_int(self.sysctl_path)
+        except TargetError:
+            if not self.target.file_exists(self.sysctl_path):
+                raise TargetError('schedstats not supported by target. '
+                                  'Ensure CONFIG_SCHEDSTATS is enabled.')
+
+        self.target.write_value(self.sysctl_path, 1)
+
+        # Check version matches
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        match = re.search(r'version ([0-9]+)', lines[0])
+        if not match or match.group(1) != '15':
+            raise TargetError(
+                'Unsupported schedstat version string: "{}"'.format(lines[0]))
+
+        # Take a sample of the schedstat file to figure out which channels to
+        # create.
+        # We'll create a site for each CPU and a site for each sched_domain.
+        for site, measures in self._get_sample().iteritems():
+            if site.startswith('cpu'):
+                measurement_category = 'schedstat_cpu'
+            else:
+                measurement_category = 'schedstat_domain'
+
+            for measurement_name in measures.keys():
+                # TODO: constructing multiple MeasurementTypes with same
+                # params. Does it matter?
+                measurement_type = MeasurementType(
+                    measurement_name, '', measurement_category)
+                self.add_channel(site=site,
+                                 name='{}_{}'.format(site, measurement_name),
+                                 measure=measurement_type)
+
+    def teardown(self):
+        self.target.write_value(self.sysctl_path, self.old_sysctl_value)
+
+    def _get_sample(self):
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        ret = OrderedDict()
+
+        # Example /proc/schedstat contents:
+        #
+        # version 15
+        # timestamp <timestamp>
+        # cpu0 <cpu fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+        # cpu1 <cpu_fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+
+        curr_cpu = None
+        for line in lines[2:]:
+            tokens = line.split()
+            if tokens[0].startswith('cpu'):
+                curr_cpu = tokens[0]
+                site = curr_cpu
+                measures = CPU_MEASURES
+            elif tokens[0].startswith('domain'):
+                if not curr_cpu:
+                    raise TargetError(
+                        'Failed to parse schedstats, found domain before CPU')
+                # We'll name the site for the domain like "cpu0domain0"
+                site = curr_cpu + tokens[0]
+                measures = DOMAIN_MEASURES
+            else:
+                raise TargetError(
+                    'Unrecognised schedstats line: "{}"'.format(line))
```

Comment:
I'd rather not design & document an API for that given that these EAS stats are probably the only ever user (also TBH we have broken schedstats with those additions, we should really change the version string to "15.eas0" or something IMO). How about if:

1. Unrecognised lines are ignored with a warning
2. I rewrite this so we can inherit from it and write a subclass that knows about the EAS stats
3. Then, if we decide that the EAS stats format is "stable", or come up with a way to fixup the version string so we can detect changes, we just add support for the EAS stats directly here